src/compiler.ts

```diff
@@ -10585,11 +10584,23 @@ export class Compiler extends DiagnosticEmitter {
     var temp = flow.getTempLocal(type);
     if (!flow.canOverflow(expr, type)) flow.setLocalFlag(temp.index, LocalFlags.WRAPPED);
     flow.setLocalFlag(temp.index, LocalFlags.NONNULL);
-    expr = module.if(
-      module.local_tee(temp.index, expr, type.isManaged),
-      module.local_get(temp.index, type.toRef()),
-      this.makeStaticAbort(this.ensureStaticString("unexpected null"), reportNode) // TODO: throw
-    );
+    if (type.kind >= TypeKind.FUNCREF && type.kind <= TypeKind.DATAREF) {
```

Comment:
Looks like this check can be collapsed to
```suggestion
    if (type.isExternalReference) {
```

Also reminds me that we should probably update `i31ref` to be nullable [here](https://github.com/AssemblyScript/assemblyscript/blob/1f641be653c9a06dba18927715a067d3c865598c/src/types.ts#L697-L701) to match the current [state of GC](https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#reference-types).

Reply:
Already done =) See recent commit

Reply:
Seems you had the same idea already :)